text/0059-functions.md

```diff
@@ -0,0 +1,282 @@
+This means we are not expanding the expressive power of Cedar in any way. 
+We are also adding more than one way to accomplish the same task.
+Brining back our SemVer example:
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+This policy does accomplish the user's goal of encoding the SemVer relationship.
+The problem here is readability. A reader of this policy has to identify that this is implementing the standard semantic version comparison operator, and not some bespoke versioning scheme. It’s also hurts maintainability, especially if you have multiple policies that reason about SemVers, you have to repeat the logic in each policy. This violates some basic software engineering tenets (Don’t Repeat Yourself), allowing bugs to sneak in via typos or botched copy/paste jobs. 
+
+### No Custom Parsing
+Using extension functions lets you implement a custom parser for your datatype. 
+This is useful even if your data type can be expressed as pure Cedar.
+For example, both the (existing) decimal expression and (proposed) semver extension make use of this:
+```
+decimal("1.1")
+semver("2.3.5")
+```
+vs
+```
+decimal(1,1)
+semver(2,3,5)
+```
+
+### Readability/Policies are no longer standalone
+A policy can no longer be read by itself, it has to be read in the context of all function definitions it uses.
+Policies that use an inordinate number of functions may be very opaque to read.
+
+## Alternatives
+
+### Type Annotations
+We could require/allow Cedar functions to have type annotations, taking type definitions from either the Schema or allowing them to be inline.
+
+Example:
+
+Schema:
+```
+type SemVer = { major : Long, minor : Long, patch : Long };
+```
+Policy:
+```
+function semver(major : Long, minor : Long, patch : Long) -> Semver {
+  { major : major, minor : minor, patch : patch }
+};
+```
+
+This would allow functions to typechecked in absence of a use-site, and allow for easier user specification of intent.
+It would also probably result in clearer type checker error messages.
+
+This introduces the following questions:
+
+1. Do we allow `type` declaration in policies, or just in schemas?
+2. Are type annotations enforced dynamically à la Contracts, or are they just ignored at runtime?
+  1. If they are dynamically enforced, that implies access to the schema to unfold type definitions. It also may introduce redundant type checking.
+3. Are type annotations required or optional?
+
+
+Leaving them out for this RFC only precludes only one future design decision, making type annotations required.
+This decision feels unlikely, as we want Cedar to be useful without using a schema/the validator.
+Adding optional type annotations in the future is backwards compatible. 
+
+### Let functions call other functions
```

Comment:
It would be impossible to forbid cycles, unless all functions are statically known upfront, which brings up the function linking and management question above.

Reply:
Even if it was, I think the readability constraint is also a reason to not allow this